### PR TITLE
Documentation iFrame Widget

### PR DESCRIPTION
**Pull request by joern-vh**
includeToComponent.js includes code, which has to be implemented inside
your component in order to receive the Information - guess you know better the right place.

This is what it looks like on Imslp.org:

image

### CODE REVIEW

**Comment by chaoran-chen** [comment]
I've just deployed the demo:
http://54.68.164.199:8083/

File: iFrame Widget/Dokumentation/index.html [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+<html lang="en">
+    <head>
+        <meta charset="utf-8">
+        <meta http-equiv="X-UA-Compatible" content="IE=edge">
+        <meta name="viewport" content="width=device-width, initial-scale=1">
+        <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
+        <title>MusicConnectionMachine Widget</title>
+
+        <!-- Bootstrap -->
+        <!--<link href="css/bootstrap.min.css" rel="stylesheet">-->
```
Comment (martomi) [inline iFrame Widget/Dokumentation/index.html]: We don't need that? You can delete it if so.

File: iFrame Widget/Dokumentation/index.html [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        
+        <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
+        <link href="style.css" rel="stylesheet">
+
+        <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
+        <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
+        <!--[if lt IE 9]>
+        <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
+        <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
+        <![endif]-->
```
Comment (martomi) [inline iFrame Widget/Dokumentation/index.html]: We probably don't need to support IE8 since React itself [dropped support](https://facebook.github.io/react/blog/2016/01/12/discontinuing-ie8-support.html)

File: iFrame Widget/Dokumentation/index.html [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                    <div class="col-sm-6">
+                        <h2>Timeline Component</h2>
+                        <p>
+                            Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
+                        </p>
+                    </div>
+                    <div class="col-sm-6">
+                        <h2>Relations Component</h2>
+                        <p>
+                            Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
```
Comment (martomi) [inline iFrame Widget/Dokumentation/index.html]: I would replace the placeholder text with a small meaningful description. Maybe something like:

> The relations component is a great way for users to explore interesting but less known facts about music pieces, composers and musicians. Users have the possibility to search and browse through all the relations and related entities of a given entity of interest.

Similar for timeline 😄

File: iFrame Widget/Dokumentation/index.html [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
...s="string">"("</span>)+1);<br>        <span class="var">var</span> <span class="varName">lastname</span> = <span class="varName">composerPart</span>.slice(0,<span class="varName">composerPart</span>.indexOf(","));<br>        <span class="var">var</span> <span class="varName">firstname</span> = <span class="varName">composerPart</span>.slice(<span class="varName">composerPart</span>.indexOf(",")+2,<span class="varName">composerPart</span>.length-1);<br>        <span class="var">var</span> <span class="varName">composer</span> = <span class="varName">firstname</span> + "_" + <span class="varName">lastname</span>;<br>        <span class="varName">urlAppendix</span> = <span class="string">"Composition:"</span> + <span class="varName">composition</span> + <span class="string">";Composer:"</span> + <span class="varName">composer</span>;<br>    }<br>    <span class="comment">//CHANGE URL to domain</span><br>    <span class="tag">document</span>.getElementById(<span class="style">"widget"</span>).src = <span class="string">"WIDGET-URL"</span> + <span class="string">"?"</span> + <span class="varName">urlAppendix</span>;<br>}<br><span class="functionName">startUpWidget()</span>;</code></pre>
```
Comment (martomi) [inline iFrame Widget/Dokumentation/index.html]: You should delete the example urls from here since they were only for testing. Maybe add a comment above the url variable for the expected format of the url with one example

File: iFrame Widget/Dokumentation/index.html [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
+        <!-- Include all compiled plugins (below), or include individual files as needed -->
+        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
+        <script>
+            function startUpWidget() {
+                //Transform URL to Information
+                //var url = window.location.href;
+                //Only for Testing!
+                var url = "http://imslp.org/wiki/Category:Beethoven,_Ludwig_van";
+                var url = "http://imslp.org/wiki/Studien_im_Generalbass,_Contrapunkt_und_in_der_Compositionslehre_(Beethoven,_Ludwig_van)";
```
Comment (martomi) [inline iFrame Widget/Dokumentation/index.html:127]: You already wrote in the example code snippet how imslp are supposed to integrate it in their website with:

`var url = window.location.href;`

so you could safely delete the commented out code from here.

Also why do you define the url variable twice? You're only using the second definition as the first one gets immediately overridden, right? Delete one of them.

**Review by martomi — CHANGES_REQUESTED** [review]
A few comments from my side. Also look at the picture below: when I jump to a section the header / title of that section is partially obscured by the floating top bar.

screen shot 2017-04-02 at 18 46 20

**Comment by vviro** [comment]
Cool, well done!

A few suggestions: 

1. how about disabling the almost always on animation? It makes the CPUs work quite a bit and doesn't add that much value to the UI - some users might get upset enough to complain about this.
2. The text overlays on the circles should probably not block the view in random places but have a fixed position outside the graph space. Otherwise it blocks further navigation and mouse-overs. What do you think?
3. Probably it should be possible to show more clearly that the displayed entities are all related to "Mozart" or whoever is the anchor. This should be clear and this entity should be presented (if at all) in a different way than the rest.
4. On the widget page the details of parsing the composers from the URL and other such minutia should not be presented - or at least packed into separate functions. The point of this documentation is readability, it is there to show how simple an embedding is. The particular details of handling variable passing on the IMSLP page are irrelevant.

**Comment by vviro** [comment]
5. Please bug the other teams for the time-related data - without it the timeline widget is useless and it might turn out to be so in the end if they can't provide you good data on time. Don't wait for them, ask them for it every day until you get it. There is a real risk that your work on it will turn out to be for nothing.

**Comment by vviro** [comment]
6. Link all the widgets you create with the widget page in the same way it's done in the Peachnote Viewer widget. This is very important. Otherwise nobody knows where this came from, who did it and so on. You want to get kudos and feedback, don't you?

File: iFrame Widget/Dokumentation/index.html (hunk 5 above)
Comment (joern-vh) [reply iFrame Widget/Dokumentation/index.html:127]: Yes, just forgot to comment the other one - was for testing whether Composer or Composition.

**Comment by joern-vh** [comment]
@vviro I added the Imslp Example because I think this is the Most Important website for the Widget - right? They just need to include this function and do not have to implement something on their own. What do you mean with 6.?

**Comment by vviro** [comment]
@Binabiko The integration must be implemented, no question about it, it just doesn't need to be presented on the widget page in form of an example.

**Comment by vviro** [comment]
@Binabiko I mean that there should be a link in the widget's frame pointing to the widget's page (this description/documentation page)

**Comment by vviro** [comment]
7. Regarding the timeline, with how many entries does the current design work? 5 events, 10 events, 20 events, 50 events? What is the expected distribution of label text lengths? 
The mockup shows the case with 5 rather short labels. Would the widget be usable with any other kind of data (more labels, longer texts)? If not, how could one still make it work? Maybe show the labels only on mouse-over? One label at a time or a neighborhood of labels? Or maybe show some labels distributed over time by default and select which labels to display based on their "popularity" that would be provided by the API? If so, would the API provide this information, is this coordinated with the data groups?

**Comment by chaoran-chen** [comment]
Hi @vviro. Sorry that I am answering so late.

1., 2., 3.: These points are regarding the bubble chart, I guess. As I have just written in [another issue](https://github.com/MusicConnectionMachine/VisualizationG5/issues/52#issuecomment-291316393), the bubble chart was only designed as an experiment / proof of concept and is not ready to use since we have decided in our meeting to not use a graphical (bubble-based) visualization at the beginning. But if you like it, we can build it as a real widget.

7. Please see the screenshots [here](https://github.com/MusicConnectionMachine/VisualizationG5/issues/52#issuecomment-291316393). I believe that with a bit visual fine-tuning (colored boxed, margins, pre-defined zooming), it is possible to display hundreds or even thousands of events with the component. Then, it is, of course, impossible to read all of the information but by using the search it shouldn't be a problem.
If a label becomes too long, I would suggest to cut it after a certain number of words and then display the remaining text in the tooltip.

**Comment by martomi** [comment]
Shall we merge that for now? I'll approve the changes so far from my side.

**Review by martomi — APPROVED** [review]
